### PR TITLE
note on step and stop for gradient descent

### PR DESCRIPTION
**Pull request by dehann**
xref https://github.com/JuliaManifolds/Manopt.jl/issues/86#issuecomment-864185971

### CODE REVIEW

**Comment by kellertuer** [comment]
I would prefer that both get demonstrated also in their effect then, not just as a note, i.e. 
* run the mean with Armijo rule and see that it is faster (less iterations) – show function value afterwards
* then increase tolerance in stopping criterion to illustrate that the cost can even be decreased if we “sharpen” the stopping criterion.

**Comment by kellertuer** [comment]
And looking at the numbers it might be better to only do that with the median, since that runs quite some iterations ;)

**Comment by kellertuer** [comment]
After line 142 in the MeanAndMedian tutorial add something like (with proper formatting Ian `#`s in front of text and such)

A way to get better performance (and convergence) is to switch the default [`ConstantStepsize`](@ref)`(1.0)` with a step size that performs better, for example the [`ArmijoLinesearch`](@ref)`()`.
' Since the mean using a constant step size is already quite fast, we even have to tweak the default values for the `contractionFactor` and the `sufficientDecrease`. We obtain

````
xMean2 = gradient_descent(
    M,
    F,
    gradF,
    data[1];
    stepsize = ArmijoLinesearch(1.0, ExponentialRetraction(), 0.99, 0.5),
    debug=[:Iteration, " | ", :x, " | ", :Change, " | ", :Cost, "\n", :Stop],
)
````

And while it finishes in already 5 steps, it is even (just so slightly) better than the previous computation

````
F(M, xMean) - F(M, xMean2)
````

To even have more precision, we can further require a smaller gradient norm. For this we can change the
[`StoppingCriterion`](@ref) used, where several criteria can be combined using `&` and/or `|`.
If we want to decrease the final gradient (from less that `1e-8`) norm but keep the maximal number of iterations to be 200, we can run

```julia
xMean3 = gradient_descent(
    M,
    F,
    gradF,
    data[1];
    stepsize = ArmijoLinesearch(1.0, ExponentialRetraction(), 0.99, 0.5),
    stopping_criterion = StopAfterIteration(200) | StopWhenGradientNormLess(1e-15),
    debug=[:Iteration, " | ", :x, " | ", :Change, " | ", :Cost, "\n", :Stop],
)
```
so this now takes 10 iterations, we get a very small gradient, but we do not gain that much anymore in the cost itself

```julia
F(M,xMean2)-F(M,xMean3)
```

**Comment by dehann** [comment]
Hi @kellertuer , oh sure I will do -- suggestions sound good.  Sorry I missed these messages yesterday.

**Comment by dehann** [comment]
I'm not 100% familiar with the Literacy.jl documentation support package but hopefully got the text added correctly.  

I copied over your text with a few small grammar updates and one additional sentence (to help avoid the problem both Affie and I stepped in):

> Note, other optimization cases may vary where adjusting default stepsize can have a more significant impact on performance.

**Comment by dehann** [comment]
i see Formatter is giving 3 warnings, best I can make out is:
```
input data has violations
```

but not sure if that is just a bad parse on functions for docs, or if that is an actual warning of interest?

**Comment by kellertuer** [comment]
Concerning the formatting I hope that the Formatter now reports missing format as changes you can just accept, if not, let me know.

File: src/tutorials/MeanAndMedian.jl [hunk 1]
```diff
@@ -140,6 +140,42 @@ xMean = gradient_descent(
...
+    F,
+    gradF,
+    data[1];
+    stepsize=ArmijoLinesearch(1.0, ExponentialRetraction(), 0.99, 0.5),
+    debug=[:Iteration, " | ", :x, " | ", :Change, " | ", :Cost, "\n", :Stop],
+)
+# which finishes in 5 steaps, just slightly better than the previous computation.
+F(M, xMean) - F(M, xMean2)
+# Note, other optimization cases may vary where adjusting default stepsize 
+# can have a more significant impact on performance.
```
Comment (kellertuer) [inline src/tutorials/MeanAndMedian.jl:159]:
```suggestion
# Note, other optimization tasks may have other speedup opportunities.
```

File: src/tutorials/MeanAndMedian.jl [hunk 2]
```diff
@@ -140,6 +140,42 @@ xMean = gradient_descent(
 )
 nothing #hide
 #
+# A way to get better performance (and convergence) is to switch the default
```
Comment (kellertuer) [inline src/tutorials/MeanAndMedian.jl:143]:
```suggestion
# A way to get better performance and for convex and coercive costs a guaranteed convergence is to switch the default 
```

**Review by kellertuer — APPROVED** [review]
